商品情報編集機能

samanosuke-akechi opened the pull request:
# What
商品情報編集機能

# Why
商品情報の編集ができるようにするため

- ログイン状態の出品者は、商品情報編集ページに遷移できる動画
https://gyazo.com/cee31f0b71c460479a1916971fc47076
- 正しく情報を記入すると、商品の情報を編集できる動画
https://gyazo.com/0610cd9f3c74c07c687b4fbab0ba0ed5
- 入力に問題のある状態では商品の情報は編集できず、エラーメッセージが出力される動画
https://gyazo.com/c4d3a7a471a379ea1584febb5e752d27
- 何も編集せずに更新をしても画像無しの商品にならない動画
https://gyazo.com/35c912e58e6c8d0c73dd093ec21e80a4
- ログイン状態のユーザーであっても、URLを直接入力して出品していない商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/62854ad8ab583bbe48ab5b3dc62aab6b
- ログアウト状態のユーザーは、URLを直接入力して商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画
https://gyazo.com/7739eef3046b6c0d7b6e3a0beb2e7e05
- 商品名やカテゴリーの情報など、すでに登録されている商品情報は商品情報編集画面を開いた時点で表示される動画（画像に関しては、表示されない状態で良い）
https://gyazo.com/6ee15eef96b8fcdd037ee32eae8dedc4

samanosuke-akechi committed:
edit,updateのルーティング

samanosuke-akechi committed:
商品情報編集ページのビューファイル配置

samanosuke-akechi committed:
商品情報編集ページへのパスを記述

samanosuke-akechi committed:
商品情報取得・ビューへの反映

samanosuke-akechi committed:
編集機能実装

保存に成功したら、商品詳細ページにリダイレクトされる。
バリデーションに引っ掛かるなどで、保存に失敗したら編集ページに止まる。

samanosuke-akechi committed:
もどるボタン実装

samanosuke-akechi committed:
エラーメッセージの表示

samanosuke-akechi committed:
出品者以外の商品編集ページへのアクセスを不許可

出品者以外が直接URLを入力してもトップページにリダイレクトされる。

TakashiKozuma commented on app/controllers/items_controller.rb:
```diff
@@ -23,6 +23,22 @@ def show
+  end
+
+  def update
+    @item = Item.find(params[:id])
```
```ruby
    @item = Item.find(params[:id])
```
23,27,34行目で定義している@itemの記述が重複しているようです。
こちらはprivate以下にメソッド化して、beforeアクションで呼び出す形に修正しておきましょう。
複数回使用する変数をメソッド化することで可読性が上がり、改修作業が発生した際も手を加えやすくなります。

TakashiKozuma commented on app/controllers/items_controller.rb:
```diff
@@ -23,6 +23,22 @@ def show
 
+  def edit
+    @item = Item.find(params[:id])
+    unless current_user.id == @item.user_id
+      redirect_to action: :index
+    end
```
```ruby
    unless current_user.id == @item.user_id
      redirect_to action: :index
    end
```
こちらの条件分岐はupdateアクションにも適用されるように修正しましょう。
通常の挙動ではupdateはできませんが、検証ツールなどを用いて不正なユーザーに商品編集されてしまうことを防ぐためです。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
動画も確認できました。
コメントを致しましたので、ご確認お願いいたします。

samanosuke-akechi committed:
リファクタリング（商品情報取得のコード）

set_itemメソッドを定義してその処理としてidに紐づく商品情報の取得を行う。
before_actionで商品詳細表示・編集・更新時にメソッドが呼び出される。

samanosuke-akechi committed:
リファクタリング（出品者以外の編集・更新の不許可）

出品者以外の編集・更新を不許可にする処理をnot_exhibitorメソッドの処理に集約。
更新のupdateアクションにも追記。

samanosuke-akechi commented:
上記、2点の修正を行いましたので、確認のほどよろしくお願いします。

kuu0123 commented on app/controllers/items_controller.rb:
```diff
@@ -20,12 +21,34 @@ def create
+  end
+
+  def edit
+    not_exhibitor
+  end
+
+  def update
+    not_exhibitor
```
```ruby
not_exhibitor
```
こちらのメソッドは処理の途中で呼び出すものではないため、予めbefore_actionで呼び出すように変更しておきましょう。

samanosuke-akechi replied:
ご指摘の通り修正しましたので、確認のほどよろしくお願いします。

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
細かな点となりますが、リファクタリングの観点から1点追加でコメントいたしましたので、ご確認お願いします！

samanosuke-akechi committed:
リファクタリング　not_exhibitorをbefore_actionに設定

kuu0123 commented:
レビュー依頼ありがとうございます！

商品編集機能、LGTMとさせていただきます！実装お疲れさまでした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切るようにしましょう。
引き続き実装頑張ってください！
